src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py

```diff
@@ -0,0 +1,357 @@
+            root_certificates=ca_1_pem,
+            private_key=client_key_1_pem,
+            certificate_chain=client_chain_1_pem)
+
+        # should work again...
+        self._do_one_shot_client_rpc(
+            True,
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+
+        # and again...
+        self._do_one_shot_client_rpc(
+            True,
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+
+        # now create the "persistent" clients
+        persistent_client_stub_A = self._create_client_stub(
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+        self._perform_rpc(persistent_client_stub_A, True)
+
+        persistent_client_stub_B = self._create_client_stub(
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+        self._perform_rpc(persistent_client_stub_B, True)
+
+        # moment of truth!! client should reject server because the
+        # server switch cert...
+        self._do_one_shot_client_rpc(
+            False,
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+
+        # now should work again...
+        self._do_one_shot_client_rpc(
+            True,
+            root_certificates=ca_2_pem,
+            private_key=client_key_1_pem,
+            certificate_chain=client_chain_1_pem)
+
+        # and again...
+        self._do_one_shot_client_rpc(
+            True,
+            root_certificates=ca_2_pem,
+            private_key=client_key_1_pem,
+            certificate_chain=client_chain_1_pem)
+
+        # client should be rejected by server if with_client_auth
+        self._do_one_shot_client_rpc(
+            not self.with_client_auth,
+            root_certificates=ca_2_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+
+        # here client should reject server...
+        self._do_one_shot_client_rpc(
+            False,
+            root_certificates=ca_1_pem,
+            private_key=client_key_2_pem,
+            certificate_chain=client_chain_2_pem)
+
+        # persistent clients should continue to work
+        self._perform_rpc(persistent_client_stub_A, True)
+        self._perform_rpc(persistent_client_stub_B, True)
+
+
+class ServerSSLCertConfigFetcherParamsChecks(unittest.TestCase):
+
+    def test_params_checks(self):
+        for initial_config, subsequent_config_fetcher, exc_msg in (
+            (None, str, 'initial certificate config must be specified'),
+            (1, '1', 'certificate config fetcher must be callable'),):
+            with self.assertRaises(ValueError) as exc:
```

Comment:
exception context no longer needed since the it was used only to check the error message, and i'm not doing that any more since as you suggest earlier it's not advisable.